core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -591,22 +594,43 @@ public PlanWithProperties visitSimpleTableExecuteNode(SimpleTableExecuteNode nod
         @Override
         public PlanWithProperties visitTableWriter(TableWriterNode node, StreamPreferredProperties parentPreferences)
         {
-            return visitTableWriter(node, node.getPartitioningScheme(), node.getSource(), parentPreferences);
+            return visitTableWriter(node, node.getPartitioningScheme(), node.getSource(), parentPreferences, node.getTarget());
         }
 
         @Override
         public PlanWithProperties visitTableExecute(TableExecuteNode node, StreamPreferredProperties parentPreferences)
         {
-            return visitTableWriter(node, node.getPartitioningScheme(), node.getSource(), parentPreferences);
+            return visitTableWriter(node, node.getPartitioningScheme(), node.getSource(), parentPreferences, node.getTarget());
         }
 
-        private PlanWithProperties visitTableWriter(PlanNode node, Optional<PartitioningScheme> partitioningSchemeOptional, PlanNode source, StreamPreferredProperties parentPreferences)
+        private PlanWithProperties visitTableWriter(
+                PlanNode node,
+                Optional<PartitioningScheme> partitioningSchemeOptional,
+                PlanNode source,
+                StreamPreferredProperties parentPreferences,
+                WriterTarget writerTarget)
         {
-            return visitPartitionedWriter(node, partitioningSchemeOptional, source, parentPreferences);
+            return visitPartitionedWriter(node, partitioningSchemeOptional, source, parentPreferences, writerTarget);
         }
 
-        private PlanWithProperties visitPartitionedWriter(PlanNode node, Optional<PartitioningScheme> optionalPartitioning, PlanNode source, StreamPreferredProperties parentPreferences)
+        private PlanWithProperties visitPartitionedWriter(PlanNode node, Optional<PartitioningScheme> optionalPartitioning, PlanNode source, StreamPreferredProperties parentPreferences, WriterTarget writerTarget)
         {
+            // TODO - Support scale task writers for partitioned tables (https://github.com/trinodb/trino/issues/13379)
+            if (optionalPartitioning.isEmpty()
+                    && isTaskScaleWritersEnabled(session)
+                    && writerTarget.supportsReportingWrittenBytes(plannerContext.getMetadata(), session)) {
+                PlanWithProperties newSource = source.accept(this, any());
```

Comment:
This should be `defaultParallelism(session)` instead of `any()`. We prefer upstream nodes are parallelised